test/app-tap/datetime.test.lua

```diff
@@ -0,0 +1,203 @@
+                  function() T:set{ sec = 61} end)
+    assert_raises(test,  range_check_error('sec', -1, {0, 60}),
+                  function() T:set{ sec = -1} end)
+
+    local only1 = 'only one of nsec, usec or msecs may defined simultaneously'
+    assert_raises(test, only1, function()
+                    T:set{ nsec = 123456, usec = 123}
+                  end)
+    assert_raises(test, only1, function()
+                    T:set{ nsec = 123456, msec = 123}
+                  end)
+    assert_raises(test, only1, function()
+                    T:set{ nsec = 123456, usec = 1234, msec = 123}
+                  end)
+end)
+
+local function invalid_tz_fmt_error(val)
+    return ('invalid time-zone format %s'):format(val)
+end
+
+test:test("Time invalid tzoffset in :set{} operations", function(test)
+    test:plan(10)
+
+    local T = date.new{}
+    local bad_strings = {
+        'bogus',
+        '0100',
+        '+-0100',
+        '+25:00',
+        '+99:00',
+        '-99:00',
+    }
+    for _, val in ipairs(bad_strings) do
+        assert_raises(test, invalid_tz_fmt_error(val),
+                      function() T:set{ tzoffset = val } end)
+    end
+
+    local bad_numbers = {
+        800,
+        -800,
+        10000,
+        -10000,
+    }
+    for _, val in ipairs(bad_numbers) do
+        assert_raises(test, range_check_error('tzoffset', val, {-720, 720}),
+                      function() T:set{ tzoffset = val } end)
+    end
+end)
+
+
+test:test("Time :set{day = -1} operations", function(test)
+    test:plan(14)
+    local tests = {
+        {{ year = 2000, month = 3, day = -1}, '2000-03-31T00:00:00Z'},
+        {{ year = 2000, month = 2, day = -1}, '2000-02-29T00:00:00Z'},
+        {{ year = 2001, month = 2, day = -1}, '2001-02-28T00:00:00Z'},
+        {{ year = 1900, month = 2, day = -1}, '1900-02-28T00:00:00Z'},
+        {{ year = 1904, month = 2, day = -1}, '1904-02-29T00:00:00Z'},
+    }
+    local T
+    for _, row in ipairs(tests) do
+        local args, str = unpack(row)
+        T = date.new(args)
+        test:is(tostring(T), str, ('checking -1 with %s'):format(str))
+    end
+    assert_raises(test, range_check_3_error(0), function() T = date.new{day = 0} end)
+    assert_raises(test, range_check_3_error(-2), function() T = date.new{day = -2} end)
+    assert_raises(test, range_check_3_error(-10), function() T = date.new{day = -10} end)
+
+    T = date.new{ year = 1904, month = 2, day = -1 }
+    test:is(tostring(T), '1904-02-29T00:00:00Z', 'base before :set{}')
+    test:is(tostring(T:set{month = 3, day = 2}), '1904-03-02T00:00:00Z', '2 March')
+    test:is(tostring(T:set{day = -1}), '1904-03-31T00:00:00Z', '31 March')
+
+    assert_raises(test, range_check_3_error(0), function() T:set{day = 0} end)
+    assert_raises(test, range_check_3_error(-2), function() T:set{day = -2} end)
+    assert_raises(test, range_check_3_error(-10), function() T:set{day = -10} end)
+end)
+
+os.exit(test:check() and 0 or 1)
```

Comment:
Are you sure you covered the entire datetime.lua with just 200 lines of tests?

Reply:
I was sure at the moment (because it was showing 82% coverage of a `datetime.lua` code liens, now, in the consolidated branch, it shows 88.2% coverage with 500 lines of code.